// @forward-software/react-native-toolbox/dist/commands/dotenv.js
"use strict";
/*
 * Copyright (c) 2020 Mattia Panzeri <mattia.panzeri93@gmail.com>
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@oclif/core");
const chalk_1 = require("chalk");
const Listr = require("listr");
const fsp = require("node:fs/promises");
const file_utils_1 = require("../utils/file-utils");
class Dotenv extends core_1.Command {
    async run() {
        const { args } = await this.parse(Dotenv);
        const sourceEnvFilePath = `./.env.${args.environmentName}`;
        const outputEnvFile = './.env';
        const sourceFilesExists = (0, file_utils_1.checkAssetFile)(sourceEnvFilePath);
        if (!sourceFilesExists) {
            this.error(`Source file ${(0, chalk_1.cyan)(sourceEnvFilePath)} not found! ${(0, chalk_1.red)('ABORTING')}`);
        }
        this.log(`Generating .env from ${(0, chalk_1.cyan)(sourceEnvFilePath)} file...`);
        const workflow = new Listr([
            {
                title: 'Remove existing .env file',
                task: async () => {
                    try {
                        await fsp.unlink(outputEnvFile);
                    }
                    catch { }
                },
            },
            {
                title: 'Generate .env file',
                task: () => fsp.copyFile(sourceEnvFilePath, outputEnvFile),
            },
        ]);
        try {
            await workflow.run();
        }
        catch (error) {
            this.error(error);
        }
    }
}
exports.default = Dotenv;
Dotenv.description = `manage .env files for react-native-dotenv
Manage .env files for react-native-dotenv for a specific environment (development, production, etc...)
`;
Dotenv.flags = {
    help: core_1.Flags.help({ char: 'h' }),
};
Dotenv.args = [
    {
        name: 'environmentName',
        description: 'name of the environment to load .dotenv file for',
        required: true,
        hidden: false,
    },
];
